sqlalchemy-stubs/future/engine.pyi

```diff
@@ -1,48 +1,98 @@
 from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import NoReturn
 from typing import Optional
+from typing import Type
+from typing import Union
 
-from .. import util as util
-from ..engine import Connection as _LegacyConnection
-from ..engine import Engine as _LegacyEngine
-from ..engine.base import OptionEngineMixin as OptionEngineMixin
+from .. import util
+from ..engine.base import Connection as _LegacyConnection
+from ..engine.base import Engine as _LegacyEngine
+from ..engine.base import NestedTransaction
+from ..engine.base import OptionEngineMixin
+from ..engine.base import Transaction
+from ..engine.create import _Debug
+from ..engine.create import _IsolationLevel
+from ..engine.create import _ParamStyle
+from ..engine.create import _ResetOnReturn
+from ..engine.url import URL
+from ..pool import Pool
 
-NO_OPTIONS: Any
+NO_OPTIONS: util.immutabledict[Any, Any] = ...
 
-def create_engine(*arg: Any, **kw: Any): ...
+def create_engine(
+    url: Union[str, URL],
+    *,
+    case_sensitive: bool = ...,
+    connect_args: Dict[Any, Any] = ...,
+    convert_unicode: bool = ...,
+    creator: Callable[..., Any] = ...,
+    echo: Union[bool, _Debug] = ...,
+    echo_pool: Union[bool, _Debug] = ...,
+    enable_from_linting: bool = ...,
+    encoding: str = ...,
+    execution_options: Dict[Any, Any] = ...,
+    hide_parameters: bool = ...,
+    implicit_returning: bool = ...,
+    isolation_level: _IsolationLevel = ...,
+    json_deserializer: Callable[..., Any] = ...,
+    json_serializer: Callable[..., Any] = ...,
+    label_length: Optional[int] = ...,
+    listeners: Any = ...,
+    logging_name: str = ...,
+    max_identifier_length: Optional[int] = ...,
+    max_overflow: int = ...,
+    module: Optional[Any] = ...,
+    paramstyle: Optional[_ParamStyle] = ...,
+    pool: Optional[Pool] = ...,
+    poolclass: Optional[Type[Pool]] = ...,
+    pool_logging_name: str = ...,
+    pool_pre_ping: bool = ...,
+    pool_size: int = ...,
+    pool_recycle: int = ...,
+    pool_reset_on_return: Optional[_ResetOnReturn] = ...,
+    pool_timeout: float = ...,
+    pool_use_lifo: bool = ...,
+    plugins: List[str] = ...,
+    query_cache_size: int = ...,
+    **kw: Any,
+) -> Engine: ...
 
 class Connection(_LegacyConnection):
-    def begin(self): ...
-    def begin_nested(self): ...
+    def begin(self) -> Transaction: ...
+    def begin_nested(self) -> NestedTransaction: ...
     def commit(self) -> None: ...
     def rollback(self) -> None: ...
     def close(self) -> None: ...
-    def execute(
+    def execute(  # type: ignore[override]
         self,
         statement: Any,
         parameters: Optional[Any] = ...,
         execution_options: Optional[Any] = ...,
-    ): ...
-    def scalar(
+    ) -> Any: ...
```

Comment:
isn't this Result-something?

Reply:
The docs seem to indicate that is the case, but the code paths get pretty complex and don't seem to always return a Result-something. I left it as `Any` for now.